core/node/base_token_adjuster/src/base_token_fetcher.rs

```diff
@@ -0,0 +1,119 @@
+use std::{fmt::Debug, num::NonZeroU64, time::Duration};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::sync::watch;
+use zksync_dal::{ConnectionPool, Core, CoreDal};
+use zksync_types::fee_model::BaseTokenConversionRatio;
+
+const CACHE_UPDATE_INTERVAL: Duration = Duration::from_millis(500);
+
+#[async_trait]
+pub trait BaseTokenFetcher: Debug + Send + Sync {
+    fn get_conversion_ratio(&self) -> BaseTokenConversionRatio;
+}
+
+#[derive(Debug, Clone)]
+pub struct DBBaseTokenFetcher {
+    pub pool: ConnectionPool<Core>,
+    pub latest_ratio: BaseTokenConversionRatio,
+}
+
+impl DBBaseTokenFetcher {
+    pub async fn new(pool: ConnectionPool<Core>) -> anyhow::Result<Self> {
+        let latest_ratio = get_latest_price(pool.clone()).await?;
+
+        // TODO(PE-129): Implement latest ratio usability logic.
+
+        tracing::debug!(
+            "Starting the base token fetcher with conversion ratio: {:?}",
+            latest_ratio
+        );
+        Ok(Self { pool, latest_ratio })
+    }
+
+    pub async fn run(&mut self, mut stop_receiver: watch::Receiver<bool>) -> anyhow::Result<()> {
+        let mut timer = tokio::time::interval(CACHE_UPDATE_INTERVAL);
+
+        while !*stop_receiver.borrow_and_update() {
+            tokio::select! {
+                _ = timer.tick() => { /* continue iterations */ }
+                _ = stop_receiver.changed() => break,
+            }
+
+            let latest_storage_ratio = get_latest_price(self.pool.clone()).await?;
+
+            // TODO(PE-129): Implement latest ratio usability logic.
+            self.latest_ratio = BaseTokenConversionRatio {
+                numerator: latest_storage_ratio.numerator,
+                denominator: latest_storage_ratio.denominator,
+            };
+        }
+
+        tracing::info!("Stop signal received, base_token_fetcher is shutting down");
+        Ok(())
+    }
+}
+
+async fn get_latest_price(pool: ConnectionPool<Core>) -> anyhow::Result<BaseTokenConversionRatio> {
```

Comment:
Can be make into a method to avoid cloning pool (and to keep structural integrity).